modules/integration/tests-integration/tests-amqp/src/test/java/org/wso2/mb/integration/tests/amqp/functional/JMSSubscriberTransactionMessageReceiveOrderTestCase.java

```diff
@@ -0,0 +1,132 @@
+    public void prepare() throws Exception {
+        super.init(TestUserMode.SUPER_TENANT_USER);
+        AndesClientUtils.sleepForInterval(15000);
+    }
+
+
+    /**
+     * 1. Start a queue receiver with transacted sessions.
+     * 2. Send 10 messages.
+     * 3. After 10 messages are received rollback session.
+     * 4. Do same 5 times. After 50 messages received commit the session and close subscriber.
+     * 5. Analyse and see if each rollback message is received in order five times.
+     *
+     * @throws org.wso2.mb.integration.common.clients.exceptions.AndesClientConfigurationException
+     * @throws javax.jms.JMSException
+     * @throws javax.naming.NamingException
+     * @throws java.io.IOException
+     * @throws CloneNotSupportedException
+     */
+    @Test(groups = {"wso2.mb", "queue", "transactions"})
+    public void performJMSSubscriberTransactionMessageReciveOrderTestCase()
+            throws AndesClientConfigurationException, JMSException, NamingException, IOException,
+                   CloneNotSupportedException, AndesClientException {
+
+        // Create a initial JMS consumer client configuration
+        AndesJMSConsumerClientConfiguration consumerConfig = new
+                AndesJMSConsumerClientConfiguration(ExchangeType.QUEUE, QUEUE_DESTINATION);
+        consumerConfig.setAcknowledgeMode(JMSAcknowledgeMode.SESSION_TRANSACTED);
+        consumerConfig.setCommitAfterEachMessageCount(EXPECTED_COUNT);
+        consumerConfig.setRollbackAfterEachMessageCount(SEND_COUNT);
+        consumerConfig.setMaximumMessagesToReceived(EXPECTED_COUNT);
+        consumerConfig.setFilePathToWriteReceivedMessages(AndesClientConstants.
+                                                                  FILE_PATH_TO_WRITE_RECEIVED_MESSAGES);
+        consumerConfig.setPrintsPerMessageCount(EXPECTED_COUNT / 10L);
+
+        // Create JMS publisher configurations
+        AndesJMSPublisherClientConfiguration publisherConfig = new
+                AndesJMSPublisherClientConfiguration(ExchangeType.QUEUE, QUEUE_DESTINATION);
+        publisherConfig.setNumberOfMessagesToSend(SEND_COUNT);
+
+        // Initialize consumer client
+        AndesClient initialConsumerClient = new AndesClient(consumerConfig, true);
+        initialConsumerClient.startClient();
+
+        // Initialize queue publisher
+        AndesClient publisherClient = new AndesClient(publisherConfig, true);
+        publisherClient.startClient();
+
+        AndesClientUtils.waitForMessagesAndShutdown(initialConsumerClient,
+                                                    AndesClientConstants.DEFAULT_RUN_TIME);
+
+        AndesClientUtils.sleepForInterval(1000);
+
+        // Check whether transacted rollback operation keeps message order while redelivering messages.
+        boolean isRollbackPreservesMessageOrder =
+                initialConsumerClient.checkIfTransactedRollbackPreservesOrder(SEND_COUNT);
+
+        AndesClientUtils.sleepForInterval(2000);
```

Comment:
Doesn't waitForMessagesAndShutdown has a sleep inside it which actually verifies all the messages that are to be received are received? If so we might not need any sleep here at all.